esi_leap/api/controllers/v1/utils.py

```diff
@@ -0,0 +1,130 @@
+#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+#    License for the specific language governing permissions and limitations
+#    under the License.
+
+from oslo_policy import policy as oslo_policy
+from oslo_utils import uuidutils
+
+from esi_leap.common import exception
+from esi_leap.common import policy
+from esi_leap.common import statuses
+from esi_leap.objects import contract
+from esi_leap.objects import offer
+from esi_leap.resource_objects import resource_object_factory as ro_factory
+
+
+def get_authorized_offer(uuid_or_name, status_filter, cdict):
+    if uuidutils.is_uuid_like(uuid_or_name):
+        o = offer.Offer.get(uuid_or_name)
+        offer_objs = []
+
+        if (status_filter and o.status == status_filter) or \
+                status_filter is None:
+            try:
+                policy.authorize('esi_leap:offer:offer_admin', cdict, cdict)
+                offer_objs.append(o)
+            except oslo_policy.PolicyNotAuthorized:
+                pass
+
+    else:
+        try:
+            policy.authorize('esi_leap:offer:offer_admin', cdict, cdict)
+            offer_objs = offer.Offer.get_all({'name': uuid_or_name,
+                                              'status': status_filter})
+
+        except oslo_policy.PolicyNotAuthorized:
+            offer_objs = offer.Offer.get_all({'name': uuid_or_name,
+                                              'project_id': cdict['project_id'],
+                                              'status': status_filter})
+
+    if len(offer_objs) == 0:
+        raise exception.OfferNotFound(offer_uuid=uuid_or_name)
+    elif len(offer_objs) > 1:
+        raise exception.OfferDuplicateName(name=uuid_or_name)
+
+    verify_resource_permission(cdict, offer_objs[0].to_dict())
+
+    return offer_objs[0]
+
+
+def verify_resource_permission(cdict, offer_dict):
+    resource_type = offer_dict.get('resource_type')
+    resource_uuid = offer_dict.get('resource_uuid')
+    resource = ro_factory.ResourceObjectFactory.get_resource_object(
+        resource_type, resource_uuid)
+    if not resource.is_resource_admin(offer_dict['project_id']):
+        policy.authorize('esi_leap:offer:offer_admin', cdict, cdict)
+
+
+def get_offer_by_uuid_or_name_contract_creation(uuid_or_name):
+    if uuidutils.is_uuid_like(uuid_or_name):
+        o = offer.Offer.get(uuid_or_name)
+        if o.status != statuses.AVAILABLE:
+            return []
+        else:
+            return [o]
+    else:
+        offer_objs = offer.Offer.get_all({'name': uuid_or_name,
+                                          'status': statuses.AVAILABLE})
+
+        if len(offer_objs) > 1:
+            raise exception.OfferDuplicateName(
+                name=uuid_or_name)
+        elif len(offer_objs) == 0:
+            raise exception.OfferNotFound(
+                offer_uuid=uuid_or_name)
+
+        return offer_objs[0]
+
+
+def get_authorized_contract(uuid_or_name, status_filters, cdict):
```

Comment:
can status_filters be an optional parameter that defaults to `[]`?

Reply:
done